server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java

```diff
@@ -45,93 +52,100 @@
-    );
-
-    if (foundFiles == null || foundFiles.isEmpty()) {
-      throw new ISE("Found no files to ingest! Check your schema.");
-    }
-    log.info ("Found files: " + foundFiles);
-
-    final LinkedList<File> files = Lists.newLinkedList(
-        foundFiles
-    );
-
-    return new FileIteratingFirehose(
-        new Iterator<LineIterator>()
-        {
-          @Override
-          public boolean hasNext()
-          {
-            return !files.isEmpty();
-          }
-
-          @Override
-          public LineIterator next()
-          {
-            try {
-              return FileUtils.lineIterator(files.poll());
-            }
-            catch (Exception e) {
-              throw Throwables.propagate(e);
-            }
-          }
-
-          @Override
-          public void remove()
-          {
-            throw new UnsupportedOperationException();
-          }
-        },
-        firehoseParser
-    );
-  }
+	private static final EmittingLogger log = new EmittingLogger(LocalFirehoseFactory.class);
+
+	private final File baseDir;
+	private final String filter;
+	private final StringInputRowParser parser;
+
+	@JsonCreator
+	public LocalFirehoseFactory(
+	    @JsonProperty("baseDir") File baseDir,
+	    @JsonProperty("filter") String filter,
+	    // Backwards compatible
+	    @JsonProperty("parser") StringInputRowParser parser)
+	{
+		this.baseDir = baseDir;
+		this.filter = filter;
+		this.parser = parser;
+	}
+
+	@JsonProperty
+	public File getBaseDir()
+	{
+		return baseDir;
+	}
+
+	@JsonProperty
+	public String getFilter()
+	{
+		return filter;
+	}
+
+	@JsonProperty
+	public StringInputRowParser getParser()
+	{
+		return parser;
+	}
+
+	@Override
+	public Firehose connect(StringInputRowParser firehoseParser) throws IOException
+	{
+		if (baseDir == null)
+		{
+			throw new IAE("baseDir is null");
+		}
+		log.info("Searching for all [%s] in and beneath [%s]", filter, baseDir.getAbsoluteFile());
+
+		Collection<File> foundFiles = FileUtils.listFiles(
+		    baseDir.getAbsoluteFile(),
+		    new WildcardFileFilter(filter),
+		    TrueFileFilter.INSTANCE);
+
+		if (foundFiles == null || foundFiles.isEmpty())
+		{
+			throw new ISE("Found no files to ingest! Check your schema.");
+		}
+		log.info("Found files: " + foundFiles);
+
+		final LinkedList<File> files = Lists.newLinkedList(
+		    foundFiles);
+
+		return new FileIteratingFirehose(
+		    new Iterator<LineIterator>()
+		    {
+			    @Override
+			    public boolean hasNext()
+			    {
+				    return !files.isEmpty();
+			    }
+
+			    @Override
+			    public LineIterator next()
+			    {
+				    final File f = files.poll();
+				    try
+				    {
+					    InputStream innerInputStream = new FileInputStream(f);
+					    final InputStream outerInputStream = f.getName().endsWith(".gz")
+		              ? CompressionUtils.gzipInputStream(innerInputStream)
+		              : innerInputStream;
+
+					    return IOUtils.lineIterator(
```

Comment:
Doesn't this lead to some performance issue? AFAIK wrapping a BufferedReader around Reader is advisable.